File: years/2022/d02.py
import os.path

testing = False
day = '02'
if testing:
    f_input = os.path.join('inputs', f'input{day}-test.txt')
else:
    f_input = os.path.join('inputs', f'input{day}.txt')

def part1(input):
    with open(input) as f:
        data = [line.strip().split() for line in f]
    score = 0
    i_win = {'X': 'C', 'Y': 'A', 'Z': 'B'}
    tie = {'X': 'A', 'Y': 'B', 'Z': 'C'}
    score_map = {'X': 1, 'Y': 2, 'Z': 3}
    
    for opp, me in data:
        if opp == tie[me]:
            score += 3
        elif opp == i_win[me]:
            score += 6
        score += score_map[me]

    return score

print('Part one: ', part1(f_input))



def part2(input):
    with open(input) as f:
        data = [line.strip().split() for line in f]
    score = 0
    i_win = {'C': 'X', 'A': 'Y', 'B': 'Z'}
    they_win = {'A': 'Z', 'B': 'X', 'C': 'Y'}
    tie = {'A': 'X', 'B': 'Y', 'C': 'Z'}
    score_map = {'X': 1, 'Y': 2, 'Z': 3}
    
    for opp, res in data:
        if res == 'X':
            me = they_win[opp]
        elif res == 'Y':
            score += 3
            me = tie[opp]
        else:
            score += 6
            me = i_win[opp]
        score += score_map[me]

    return score

print('Part two: ', part2(f_input))

File: years/2024/d02.rb
# frozen_string_literal: true

require_relative 'base'

class Day < Base
  def part_1
    data = File.read(input).split("\n")
    data.map do |line|
      levels = line.split.map(&:to_i)
      report_safe?(levels) ? 1 : 0
    end.sum
  end

  def part_2
    data = File.read(input).split("\n")
    data.map do |line|
      levels = line.split.map(&:to_i)
      report_safe?(levels) || report_safe_with_dampner?(levels) ? 1 : 0
    end.sum
  end

  private

  def report_safe?(report)
    diffs = report[...-1].zip(report[1...]).map { |a, b| b - a }
    return false unless diffs.all?(&:positive?) || diffs.all?(&:negative?)
    return false unless diffs.map(&:abs).all? { |diff| 1 <= diff && diff <= 3 }

    true
  end

  def report_safe_with_dampner?(report)
    possible_reports = report.combination(report.size - 1)
    possible_reports.any? { |report| report_safe?(report) }
  end

  TESTING = false
  DAY = '02'

  def input
    TESTING ? File.join('inputs', "input#{DAY}-test.txt") : File.join('inputs', "input#{DAY}.txt")
  end
end

Day.new.run_part_1
Day.new.run_part_2
# Day.new.benchmark
